### PR TITLE
feat: post content translation

### LINKED ISSUE #162: Implement post content translation via API like Google Translate API
Currently this website does not support language translation. Because of that it tough for other user who are not familiar with chinese language to understand the content.

I like to work on this by add google translation api.

### Update:

See Twitter: https://twitter.com/bts_bighit/status/1403641483832422405

image

image

We may provide a translate button for each post, we will also provide a global switch to translate all contents shown.

### PR DESCRIPTION
**Pull request by WindSpiritSR**
Fix: https://github.com/casbin/casnode/issues/162

Deployed at: https://forum.leviatan.cn/t/1

Signed-off-by: WindSpirit simon343riley@gmail.com

### CODE REVIEW

**Comment by hsluoyz** [comment]
@kocoler @Kininaru plz review.

**Comment by hsluoyz** [comment]
@WindSpiritSR plz fix conflicts

**Comment by kocoler** [comment]
@hsluoyz I'm thinking about whether the translation function is better on the `server-side` or the `client-side`?

**Comment by hsluoyz** [comment]
@kocoler I think Twitter is client-side

**Comment by WindSpiritSR** [comment]
@hsluoyz @kocoler  I think the translation function on the `client-side` may be called maliciously, `server-side` can only translate the content in the database, and handle the code blocks in markdown better.

**Comment by WindSpiritSR** [comment]
@hsluoyz Twitter is server-side

image

File: controllers/topic.go [hunk 1]
```diff
@@ -637,6 +642,96 @@ func (c *ApiController) EditContent() {
 	c.ServeJSON()
 }
 
+func (c *ApiController) TranslTopic() {
```
Comment (hsluoyz) [inline controllers/topic.go]: Why don't use "Translate" everywhere?

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: controllers/topic.go (hunk 1 above)
Comment (WindSpiritSR) [reply controllers/topic.go]: it's too long

Comment (hsluoyz) [reply controllers/topic.go]: I think that's OK. `Transl` doesn't look good.

Comment (WindSpiritSR) [reply controllers/topic.go]: fixed

**Review by WindSpiritSR — COMMENTED** [review]
Renamed `transl` to `translate`

File: controllers/topic.go [hunk 2]
```diff
@@ -17,6 +17,11 @@ package controllers
 import (
 	"encoding/base64"
 	"encoding/json"
+	beego "github.com/beego/beego/v2/adapter"
```
Comment (hsluoyz) [inline controllers/topic.go:20]: Sort it

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Pull request merged**